Add the concept of runtime_cast to type index as an emulation of dynamic_cast..

cdglove opened the pull request:
I sent an email to the boost mailing list about this and received little response.

http://thread.gmane.org/gmane.comp.lib.boost.devel/268139

Is there an interest in adding this to typeindex, or maybe it should be a seperate library? For me its incredibly useful.

apolukhin commented on include/boost/type_index/runtime_cast.hpp:
```diff
@@ -0,0 +1,141 @@
+
+} // namespace detail
+
+#define BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI                                                                \
```
Can we remove this macro and use only the BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI_BASES macro?

cdglove replied:
Yes, I think I can get this down to a single macro when I remove the variadic template. Do you recommend also not relying on variadic macros? If so, I'll need to provide two implementations (one with variadic macros and one with BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI_BASES0, BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI_BASES1, BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI_BASES2, etc).

apolukhin commented on include/boost/type_index/runtime_cast.hpp:
```diff
@@ -0,0 +1,141 @@
+    template<typename T, typename U>
+    T runtime_cast(U& u) {
+        typedef typename std::remove_reference<T>::type impl_type;
+        impl_type* value = detail::runtime_cast_impl<impl_type>(&u, std::is_same<T, U>());
```
It's better to use boost::is_same, because it works even on antique compilers and c++98 compatible standard libraries

cdglove replied:
OK

apolukhin commented on include/boost/type_index/runtime_cast.hpp:
```diff
@@ -0,0 +1,141 @@
+#include <boost/type_index.hpp>
+#include <boost/throw_exception.hpp>
+#include <type_traits>
+#include <typeinfo>
```
Do we really need <typeinfo> header?

cdglove replied:
It's for std::bad_cast which I need for the reference versions (since they can't return nullptr). An alternative is to use a different exception type or move the reference versions to a different header.

apolukhin commented on include/boost/type_index/runtime_cast.hpp:
```diff
@@ -0,0 +1,141 @@
+};
+
+template<typename T, typename U>
+T* runtime_cast_impl(U* u, std::true_type) {
```
`std::true_type` and `std::false_type` may not be available at all the compilers. It's better to use Boost's alternatives

cdglove replied:
OK

apolukhin commented on include/boost/type_index/runtime_cast.hpp:
```diff
@@ -0,0 +1,141 @@
+    template<typename T, typename U>
+    T runtime_cast(U const& u) {
+        typedef typename std::remove_reference<T>::type impl_type;
+        impl_type* value = detail::runtime_cast_impl<impl_type>(&u, std::is_same<T, U>());
```
It may be profitable to use `is_base_and_derived` here too. In that case upcasts will cost nothing.

cdglove replied:
Good call. Thanks!

apolukhin commented:
Looks interesting! Great idea.

I've added a few comments. I'm short on time right now, so if you wish to fix those issues - you're welcome :) Otherwise, I'll fix them by myself somewhere around October.

Also we'll need docs and reduce C++14 and C++11 usage (get rid of `decltype` and variadic templates if possible).

cdglove commented:
I can make the changes and write the docs. This pull request was mostly to gauge interest and since you seem to be interested, I will take care of it. 

One thing I haven't included are overloads of runtime_pointer_cast for the smart pointer types. I think these should be included but packaged in separate headers. Thoughts?

cdglove commented on test/runtime_cast_test.cpp:
```diff
@@ -0,0 +1,239 @@
+        std::string name;
+
+struct base {
+    BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI
```
I can think of a few ways to get rid of this macro, and only have one, but to do so would, I think, require the user listing the name of the class something like;
`struct derived : base1, base2 {
  BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI((derived)(base1)(base2))
}`
I'm not sure if that's preferable to having 2 macros or not.

cdglove commented:
I've changed the code, removed the std::type_traits and the variadic template and replaced it with a version using BOOST_PP. I don't like dragging in the preprocessor lib, it seems a little heavy for this, but my macro skills are not really up to the task otherwise ;-).

cdglove commented:
I've added an initial version of some docs. There are a few things that came out of writing the docs that I I don't like with the current implementation.

i. I don't like having 2 macros, BOOST_TYPE_INDEX_REGISTER_CLASS and BOOST_TYPE_INDEX_REGISTER_RUNTIME_CLASS. I feel like there should only be one, or they can remain separate but there should be a master macro that does both by default. Or maybe BOOST_TYPE_INDEX_REGISTER_CLASS_EX that takes the base class list.

ii. I don't like that to use BOOST_TYPE_INDEX_REGISTER_RUNTIME_CLASS requires an additional include. I did this because it drags in boost/preprocessor/seq/for_each.hpp and I want to keep that out of the primary header.

iii. Actually, I'd like to find a solution that doesn't depend on BOOST_PP. I don't love the need for the ((base1)(base2)) syntax that PP sequences require.

I'd appreciate any suggestions on how to move forward.  What are your thoughts on using variadic macros?

apolukhin commented on include/boost/type_index/runtime_cast/reference_cast.hpp:
```diff
@@ -0,0 +1,58 @@
+template<typename T, typename U>
+T& runtime_cast(U& u) {
+    typedef typename boost::remove_reference<T>::type impl_type;
+    impl_type* value = detail::runtime_cast_impl<impl_type>(&u, boost::is_base_and_derived<T, U>());
```
use boost::addressof(u) instead of &u

cdglove replied:
Right. Thanks!

apolukhin commented on include/boost/type_index/runtime_cast/reference_cast.hpp:
```diff
@@ -0,0 +1,58 @@
+#include <boost/throw_exception.hpp>
+#include <boost/type_traits/remove_reference.hpp>
+#include <boost/type_traits/is_base_and_derived.hpp>
+#include <typeinfo>
```
I'm not sure that <typeinfo> header is available on all platforms when RTTI is off and I'm sure that std::bad_cast() is not available when defined BOOST_NO_RTTI. How about making a boost::bad_runtime_cast exception that derived from std::exception?

cdglove replied:
Yes, I think you are right -- I'll create a new exception class. Question: Should the throw be using the BOOST_THROW_EXCEPTION_MACRO or the boost::throw_exception function?

apolukhin commented on test/type_index_runtime_cast_test.cpp:
```diff
@@ -0,0 +1,268 @@
+    using namespace boost::typeindex;
+    base b;
+    base* b2 = runtime_pointer_cast<base>(&b);
+    BOOST_TEST_NE(b2, (base*)nullptr);
```
nullptr is not available on pre C++11 compilers. better to use 0 or NULL

cdglove replied:
OK. For some reason I thought this would be OK in tests, but of course we want the test suite to also run on older compilers so I'll fix this up.

apolukhin commented on include/boost/type_index/runtime_cast/detail/runtime_cast_impl.hpp:
```diff
@@ -0,0 +1,57 @@
+/// functionality on platorms that don't provide it or should the user
+/// desire opt in functionality instead of enabling it system wide.
+
+#include <boost/type_index.hpp>
```
boost/type_index.hpp must include boost/type_index/runtime_cast.hpp + there may be a problem with circular inclusions because most of the boost/type_index/runtime_cast/*.hpp headers include boost/type_index.hpp

cdglove replied:
Currently, boost/type_index.hpp is not including runtime_cast.hpp. If a user is using type_index wants the runtime_cast functionality, I expect that user to include runtime_cast.hpp separately. I personally like finer grained headers like this, but I realize it can be a pain and some people don't like it so I'm happy to change it around so that type_index brings in the basic runtime_cast functionality as well. It's really not a lot of code, more on that below.

apolukhin commented on include/boost/type_index/runtime_cast/reference_cast.hpp:
```diff
@@ -0,0 +1,58 @@
+template<typename T, typename U>
+T const& runtime_cast(U const& u) {
+    typedef typename boost::remove_reference<T>::type impl_type;
+    impl_type* value = detail::runtime_cast_impl<impl_type>(&u, boost::is_base_and_derived<T, U>());
```
boost::addressof(u)

cdglove replied:
OK.

apolukhin commented:
> it drags in boost/preprocessor/seq/for_each.hpp and I want to keep that out of the primary header.

That would be good, but I'm not sure how to do it better. I've made a few experiments with `template <class B0 = na, B1 = na, B2 = na, B3 = na> class bases` but that does not seem good. So looks like preprocessor library is OK here.

apolukhin commented:
I've added a few issues to fix. Fix those and I'll merge the pull request.

cdglove commented:
> That would be good, but I'm not sure how to do it better.

At some point I wrote a tool to analyze include file size. Though not a perfect indicator of compile time impacts, I find it to be a reasonably good precursor. When I run that tool on type_index with the preprocessor library here, it doesn't look too bad; only 140kb of additional code.

![pasted2](https://cloud.githubusercontent.com/assets/1856042/17700273/1e5b2df2-6393-11e6-9ad6-0ef6eb988de6.png)

So, I think this is an OK solution. It also means I think we can reorganize the header such that runtime_cast.hpp is unnecessary and type_index.hpp just includes the basic bits from the runtime_cast directory (essentially everything except the smart ptr overloads).
